File: olap/drill/drill-yarn/src/test/java/org/apache/drill/yarn/client/TestClient.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.drill.yarn.client;

import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import java.io.UnsupportedEncodingException;

import org.apache.drill.test.BaseTest;
import org.junit.Test;

public class TestClient extends BaseTest {

  /**
   * Unchecked exception to allow capturing "exit" events without actually
   * exiting.
   */

  public static class SimulatedExitException extends RuntimeException {
    private static final long serialVersionUID = 1L;
    public int exitCode;

    public SimulatedExitException(int exitCode) {
      this.exitCode = exitCode;
    }
  }

  public static class TestContext extends ClientContext {
    public static ByteArrayOutputStream captureOut = new ByteArrayOutputStream();
    public static ByteArrayOutputStream captureErr = new ByteArrayOutputStream();

    public static void testInit() {
      init(new TestContext());
      resetOutput();
    }

    @Override
    public void exit(int exitCode) {
      throw new SimulatedExitException(exitCode);
    }

    public static void resetOutput() {
      try {
        out.flush();
        captureOut.reset();
        out = new PrintStream(captureOut, true, "UTF-8");
        err.flush();
        captureErr.reset();
        err = new PrintStream(captureErr, true, "UTF-8");
      } catch (UnsupportedEncodingException e) {
        throw new IllegalStateException(e);
      }
    }

    public static String getOut() {
      out.flush();
      try {
        return captureOut.toString("UTF-8");
      } catch (UnsupportedEncodingException e) {
        throw new IllegalStateException(e);
      }
    }

    public static String getErr() {
      out.flush();
      try {
        return captureErr.toString("UTF-8");
      } catch (UnsupportedEncodingException e) {
        throw new IllegalStateException(e);
      }
    }
  }

  /**
   * Test the basics of the DrillOnYarn app. Does not try any real commands, but
   * does check for the basic error conditions.
   */

  @Test
  public void testBasics() {
    TestContext.testInit();

    // No arguments provided.

    try {
      DrillOnYarn.run(new String[] {});
      fail();
    } catch (SimulatedExitException e) {
      assert (e.exitCode == -1);
      assertTrue(TestContext.getOut().contains("Usage: "));
      TestContext.resetOutput();
    }

    // Bogus command

    try {
      DrillOnYarn.run(new String[] { "bogus" });
      fail();
    } catch (SimulatedExitException e) {
      assert (e.exitCode == -1);
      assertTrue(TestContext.getOut().contains("Usage: "));
      TestContext.resetOutput();
    }
  }

  // The idea here is to set up a simulated client environment, then
  // test each command. This is a big project.

}

File: olap/drill/contrib/native/client/test/ssl/testClient.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <cstdlib>
#include <iostream>
#include <boost/bind.hpp>
#include <boost/asio.hpp>
#include <boost/asio/ssl.hpp>

enum { max_length = 1024 };

class client
{
    public:
        client(boost::asio::io_service& io_service, boost::asio::ssl::context& context,
                boost::asio::ip::tcp::resolver::iterator endpoint_iterator)
            : socket_(io_service, context)
        {
            boost::asio::ip::tcp::endpoint endpoint = *endpoint_iterator;
            socket_.lowest_layer().async_connect(endpoint,
                    boost::bind(&client::handle_connect, this,
                        boost::asio::placeholders::error, ++endpoint_iterator));
        }

        void handle_connect(const boost::system::error_code& error,
                boost::asio::ip::tcp::resolver::iterator endpoint_iterator)
        {
            if (!error)
            {
                socket_.async_handshake(boost::asio::ssl::stream_base::client,
                        boost::bind(&client::handle_handshake, this,
                            boost::asio::placeholders::error));
            }
            else if (endpoint_iterator != boost::asio::ip::tcp::resolver::iterator())
            {
                socket_.lowest_layer().close();
                boost::asio::ip::tcp::endpoint endpoint = *endpoint_iterator;
                socket_.lowest_layer().async_connect(endpoint,
                        boost::bind(&client::handle_connect, this,
                            boost::asio::placeholders::error, ++endpoint_iterator));
            }
            else
            {
                std::cout << "Connect failed: " << error << "\n";
            }
        }

        void handle_handshake(const boost::system::error_code& error)
        {
            if (!error)
            {
                std::cout << "Enter message: ";
                std::cin.getline(request_, max_length);
                size_t request_length = strlen(request_);

                boost::asio::async_write(socket_,
                        boost::asio::buffer(request_, request_length),
                        boost::bind(&client::handle_write, this,
                            boost::asio::placeholders::error,
                            boost::asio::placeholders::bytes_transferred));
            }
            else
            {
                std::cout << "Handshake failed: " << error << "\n";
            }
        }

        void handle_write(const boost::system::error_code& error,
                size_t bytes_transferred)
        {
            if (!error)
            {
                boost::asio::async_read(socket_,
                        boost::asio::buffer(reply_, bytes_transferred),
                        boost::bind(&client::handle_read, this,
                            boost::asio::placeholders::error,
                            boost::asio::placeholders::bytes_transferred));
            }
            else
            {
                std::cout << "Write failed: " << error << "\n";
            }
        }

        void handle_read(const boost::system::error_code& error,
                size_t bytes_transferred)
        {
            if (!error)
            {
                std::cout << "Reply: ";
                std::cout.write(reply_, bytes_transferred);
                std::cout << "\n";
            }
            else
            {
                std::cout << "Read failed: " << error << "\n";
            }
        }

    private:
        boost::asio::ssl::stream<boost::asio::ip::tcp::socket> socket_;
        char request_[max_length];
        char reply_[max_length];
};

int main(int argc, char* argv[])
{
    try
    {
        if (argc != 3)
        {
            std::cerr << "Usage: client <host> <port>\n";
            return 1;
        }

        boost::asio::io_service io_service;

        boost::asio::ip::tcp::resolver resolver(io_service);
        boost::asio::ip::tcp::resolver::query query(argv[1], argv[2]);
        boost::asio::ip::tcp::resolver::iterator iterator = resolver.resolve(query);

        boost::asio::ssl::context ctx(boost::asio::ssl::context::sslv23);
        ctx.set_verify_mode(boost::asio::ssl::context::verify_peer);
        ctx.load_verify_file("../../../test/ssl/drillTestCert.pem");

        client c(io_service, ctx, iterator);

        io_service.run();
    }
    catch (std::exception& e)
    {
        std::cerr << "Exception: " << e.what() << "\n";
    }

    return 0;
}
